### ISSUE TITLE
Support resetting the vote when no new iteration is pushed but the voting rule no longer applies

### ISSUE THREAD

**Issue by outofthecave** [juharris/code-reviewer#51]
Consider a rule that only checks the PR title or description (or other fields that can be updated without pushing a new iteration to a PR) and votes on the PR. Once the vote has been cast on the PR and the PR author has updated the relevant field (for example, the description), it would be good if the Code Reviewer could automatically reset its vote.  It currently doesn't do so (at least not consistently).

I have prepared a unit test to reproduce my scenario in the function `test_review_pr_vote_based_on_description_reset` in [code-reviewer/tests/test_run.py at mazepf/reset-vote-after-desc-change · outofthecave/code-reviewer](https://github.com/outofthecave/code-reviewer/blob/mazepf/reset-vote-after-desc-change/tests/test_run.py).  (I'm not linking to an exact line number because I'm still working on that file and the line numbers may change.)

This code block is responsible for resetting the vote, but it only does so if another iteration has been pushed to the PR since the vote was cast: [code-reviewer/src/run.py at 76d5b859482f25021ff552c6a05474d309dd0202 · juharris/code-reviewer](https://github.com/juharris/code-reviewer/blob/76d5b859482f25021ff552c6a05474d309dd0202/src/run.py#L405-L422).  This is because the Code Reviewer can't know why a certain vote was cast on a PR and whether it was cast by the Code Reviewer itself or by a human (or even another automated tool) using the same Azure DevOps account.

Another difficulty is that it's not possible to determine whether a field like the PR description has been changed because no record of previous versions is being kept.  It's also not possible to determine whether a rule would have applied previously to such a field.

**Comment by juharris** [comment]
Thanks for sharing this. We discussed this idea; I'll just summarize some points and concerns here. I certainly see the value of this, especially when the votes only come from one source which is a bot, but in general I have a few concerns and I'm not sure if anything **needs** to change.

Yeah it only resets the vote for code changes. This wasn't really a conscious decision at first, since the rules were mainly about code changes at first, so an update is needed, but now I like how this works. The reason is more about when a person uses the script, which was the original intention of this tool. I might manually reject because of a non-automated comment. So, the script can't know why I voted what I voted. It doesn't know that I rejected for some other reason, so we can't reset the vote just because someone resolved a manually created comment without addressing it and then changed the title.

One idea that wouldn't work now: add a bunch of new rules at the beginning like: if the PR description is good, then `vote: NONE`. This won't work cause vote only votes when the new vote is more rejecting. Also, this will be annoying with extra notifications and it's not robust because there could be an error after the vote resets and before a reject could happen.

Again, this tool was made to make reviewing easier for me, not necessarily to help a submitter. The two's goals aren't always aligned.
 
So yeah, I see that this is an issue when the agent voting is a bot and there's only one source of votes. A separate setting to reset when there were no rejections could help, but with the JSON matcher checks, someone could have a rule that only votes when the bot isn't already rejecting, so that would be tricky. The easiest way would be to always reset the vote in the first rule, but that will be noisy for people and could yield issues if the script fails partway through running rules.
 
Oh right, finally, I don't really care about this case about issues in titles or descriptions. People should read the repo's guidelines and do things properly the first time and this will help them get it right the first time in the future. They'll make the mistake once, get annoyed, and hopefully not make the mistake again later. These mistakes are disturbing to other team members. I would rather make things annoying for 1 person for 1 PR, than to constantly annoy other team members with misleading titles and vague or empty descriptions until someone fixes the issue.

In summary, I don't see a proper and robust way to solve this. Sure, we can add an option to make things better in some cases, but I would rather that people follow guidelines and I'm also fine with how things are now.

**Comment by outofthecave** [comment]
I think I may have found a way to make it work.  I introduced a new config field `reset_votes_if_no_rule_votes` that allows resetting the Code Reviewer's vote if none of the rules vote on the PR.  Here is the implementation: https://github.com/outofthecave/code-reviewer/commit/7cd6fc348a648674b2325681c65e481ec1f01204.

I made it a new config field because I think you're right that the goals for running the Code Reviewer as a human vs. it running independently as a bot really don't align here.  So, I thought it would be best not to interfere with how it currently works, but create a separate config field to support this new scenario.  Plus, it would have been confusing to change the behaviour of `reset_votes_after_changes`, given that "after changes" is in the name.

This implementation makes the test case for my scenario pass, but I still want to add a few more test cases for other scenarios to make sure this isn't going to cause excessive notifications.  (I think it won't, but I want to be sure.)

Let me know what you think about the implementation and the name of the field!

**Comment by outofthecave** [comment]
Turns out there were a few bugs in my implementation, but now I've added more tests and ironed out the bugs: #52.  Let me know what you think!
